File: clients/python/unary_client.py
#!/usr/bin/env python3
"""
Python Unary client implementation demonstrating the "painful" polling approach.
"""

import sys
import time
import random
import argparse
import grpc

import tourney_unary_pb2
import tourney_unary_pb2_grpc


class UnaryClient:
    def __init__(self, host='localhost', port=9000, language_name='Python-3.12', prng_algorithm='random.Random'):
        self.channel = grpc.insecure_channel(f'{host}:{port}')
        self.stub = tourney_unary_pb2_grpc.UnaryArenaStub(self.channel)
        self.language_name = language_name
        self.prng_algorithm = prng_algorithm
        self.random = random.Random()
    
    def close(self):
        self.channel.close()
    
    def play(self):
        print(f"Unary Client starting: {self.language_name} ({self.prng_algorithm})")
        
        # Step 1: Register
        reg_response = self.stub.Register(
            tourney_unary_pb2.RegistrationRequest(
                language_name=self.language_name,
                prng_algorithm=self.prng_algorithm
            )
        )
        
        match_id = reg_response.match_id
        print(f"Registered with matchId: {match_id}, Status: {reg_response.status}")
        
        # Wait for opponent if needed
        if reg_response.status == "WAITING_FOR_OPPONENT":
            print("Waiting for opponent...")
            time.sleep(2)
        
        print("Starting match...")
        
        # Play rounds
        for round_num in range(1, 1001):
            # Step 2: Submit move
            move = self.random.randint(0, 2)  # 0=Rock, 1=Paper, 2=Scissors
            
            move_response = self.stub.SubmitMove(
                tourney_unary_pb2.MoveRequest(
                    match_id=match_id,
                    round_number=round_num,
                    move=move
                )
            )
            
            if move_response.status == "GAME_OVER":
                print("Game is over")
                break
            
            if move_response.status != "ACCEPTED":
                print(f"Move not accepted: {move_response.status}")
                continue
            
            # Step 3: Poll for result (THE PAINFUL PART)
            result = None
            poll_attempts = 0
            while result is None or result.status == "PENDING":
                poll_attempts += 1
                result = self.stub.CheckRoundResult(
                    tourney_unary_pb2.ResultRequest(
                        match_id=match_id,
                        round_number=round_num
                    )
                )
                
                if result.status == "PENDING":
                    time.sleep(0.01)  # Polling delay
            
            if round_num % 100 == 0:
                print(f"Round {round_num}: {result.outcome} (Polls: {poll_attempts})")
        
        print("Match completed!")


def main():
    parser = argparse.ArgumentParser(description='Python Unary Client for Paper-Rock-Scissors Arena')
    parser.add_argument('--host', default='localhost', help='Arena server host')
    parser.add_argument('--port', type=int, default=9000, help='Arena server port')
    parser.add_argument('--language', default='Python-3.12', help='Language name')
    parser.add_argument('--prng', default='random.Random', help='PRNG algorithm')
    
    args = parser.parse_args()
    
    client = UnaryClient(
        host=args.host,
        port=args.port,
        language_name=args.language,
        prng_algorithm=args.prng
    )
    
    try:
        client.play()
    finally:
        client.close()


if __name__ == '__main__':
    main()

File: clients/go/unary_client.go
package main

import (
	"context"
	"flag"
	"fmt"
	"log"
	"math/rand"
	"time"

	pb "github.com/ai-pipestream/paper-rock-scissors/clients/go/pb/tourney_unary"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

// UnaryClient implements the "painful" polling approach
type UnaryClient struct {
	conn         *grpc.ClientConn
	client       pb.UnaryArenaClient
	languageName string
	prngAlgo     string
	random       *rand.Rand
}

// NewUnaryClient creates a new unary client
func NewUnaryClient(host string, port int, languageName, prngAlgo string) (*UnaryClient, error) {
	address := fmt.Sprintf("%s:%d", host, port)
	conn, err := grpc.Dial(address, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, fmt.Errorf("failed to connect: %v", err)
	}

	return &UnaryClient{
		conn:         conn,
		client:       pb.NewUnaryArenaClient(conn),
		languageName: languageName,
		prngAlgo:     prngAlgo,
		random:       rand.New(rand.NewSource(time.Now().UnixNano())),
	}, nil
}

// Close closes the connection
func (c *UnaryClient) Close() {
	if c.conn != nil {
		c.conn.Close()
	}
}

// Play starts the match
func (c *UnaryClient) Play() error {
	ctx := context.Background()
	log.Printf("Unary Client starting: %s (%s)", c.languageName, c.prngAlgo)

	// Step 1: Register
	regResp, err := c.client.Register(ctx, &pb.RegistrationRequest{
		LanguageName:  c.languageName,
		PrngAlgorithm: c.prngAlgo,
	})
	if err != nil {
		return fmt.Errorf("registration failed: %v", err)
	}

	matchID := regResp.GetMatchId()
	log.Printf("Registered with matchId: %s, Status: %s", matchID, regResp.GetStatus())

	// Wait for opponent if needed
	if regResp.GetStatus() == "WAITING_FOR_OPPONENT" {
		log.Println("Waiting for opponent...")
		time.Sleep(2 * time.Second)
	}

	log.Println("Starting match...")

	// Play rounds
	for round := int32(1); round <= 1000; round++ {
		// Step 2: Submit move
		move := int32(c.random.Intn(3)) // 0=Rock, 1=Paper, 2=Scissors

		moveResp, err := c.client.SubmitMove(ctx, &pb.MoveRequest{
			MatchId:     matchID,
			RoundNumber: round,
			Move:        move,
		})
		if err != nil {
			return fmt.Errorf("submit move failed: %v", err)
		}

		if moveResp.GetStatus() == "GAME_OVER" {
			log.Println("Game is over")
			break
		}

		if moveResp.GetStatus() != "ACCEPTED" {
			log.Printf("Move not accepted: %s", moveResp.GetStatus())
			continue
		}

		// Step 3: Poll for result (THE PAINFUL PART)
		pollAttempts := 0
		var result *pb.ResultResponse
		for {
			pollAttempts++
			result, err = c.client.CheckRoundResult(ctx, &pb.ResultRequest{
				MatchId:     matchID,
				RoundNumber: round,
			})
			if err != nil {
				return fmt.Errorf("check result failed: %v", err)
			}

			if result.GetStatus() != "PENDING" {
				break
			}

			time.Sleep(10 * time.Millisecond) // Polling delay
		}

		if round%100 == 0 {
			log.Printf("Round %d: %s (Polls: %d)", round, result.GetOutcome(), pollAttempts)
		}
	}

	log.Println("Match completed!")
	return nil
}

func main() {
	host := flag.String("host", "localhost", "Arena server host")
	port := flag.Int("port", 9000, "Arena server port")
	language := flag.String("language", "Go-1.21", "Language name")
	prng := flag.String("prng", "math/rand", "PRNG algorithm")
	flag.Parse()

	client, err := NewUnaryClient(*host, *port, *language, *prng)
	if err != nil {
		log.Fatalf("Failed to create client: %v", err)
	}
	defer client.Close()

	if err := client.Play(); err != nil {
		log.Fatalf("Play failed: %v", err)
	}
}
